lib/k8s/transport.rb

```diff
@@ -144,7 +145,10 @@ def build_excon
     # @param path [Array<String>] join path parts together to build the full URL
     # @return [String]
     def path(*path)
-      File.join('/', *path)
+      path_components = ['/', *path]
+      # Only add the path_prefix if it isn't there already...
+      path_components.insert(0, path_prefix) unless [*path].join.match?(/^#{path_prefix}/)
```

Comment:
`[*path]` is redundant, it will just return a copy of the same array:

```
> path = ['a', 'b']
> [*path] 
  => ['a', 'b']
```

`[*path].join.match` will not add slashes to the path:

```
['a', 'b', 'c'].join => 'abc'
['a', 'b', 'c'].join('/') => 'a/b/c'
```

Also the regex is unnecessary, it could just be `path.join('/').start_with?(path_prefix)`

Maybe something like this:

```ruby
def path(*path)
  result = path.join('/')
  result = File.join(path_prefix, path) unless result.start_with?(path_prefix)
  result.start_with?('/') ? result : "/#{result}"
end
```

But is it actually necessary to skip adding the prefix if path already includes it?

That would force you to use a prefix that is not a kube api endpoint path. If the prefix is `/api`, maybe it should just generate `/api/api`. If you don't want that, just don't add the prefix to your server uri.

Reply:
That being said, I think the whole method can then be:

```
def path(*path)
  File.join(path_prefix, *path)
end
```

if the `path_prefix` is set to `/` in the initializer in case the uri path was empty.

Reply:
Using this one-line `path` yielded what looks to be a recursive issue with the construction of the path containing a prefix:

```
2.5.3 :001 > client = K8s::Client.config(K8s::Config.load_file(ENV['KUBECONFIG']))
 => #<K8s::Client:0x00007fd2e01d15b8 @transport=#<K8s::Transport:0x00007fd2e01d1ae0 @server="https://REDACTED.example.com:443", @path_prefix="/k8s/clusters/c-vkf6c", @auth_token="...
2.5.3 :002 > client.api.resource('pods', namespace: 'cattle-system').list.each do |pod|
2.5.3 :003 >   puts "namespace=#{pod.metadata.namespace} pod: #{pod.metadata.name}"
2.5.3 :004?> end
Traceback (most recent call last):
       16: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
       15: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
       14: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
       13: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/cli.rb:463:in `exec'
       12: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/cli/exec.rb:28:in `run'
       11: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/cli/exec.rb:74:in `kernel_load'
       10: from /Users/jlg438/.rvm/gems/ruby-2.5.3/gems/bundler-1.17.3/lib/bundler/cli/exec.rb:74:in `load'
        9: from exe/console:8:in `<top (required)>'
        8: from (irb):2
        7: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/api_client.rb:73:in `resource'
        6: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/api_client.rb:61:in `find_api_resource'
        5: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/api_client.rb:54:in `api_resources'
        4: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/api_client.rb:46:in `api_resources!'
        3: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/transport.rb:326:in `get'
        2: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/transport.rb:250:in `request'
        1: from /Users/jlg438/Development/ruby/k8s-client/lib/k8s/transport.rb:228:in `parse_response'
K8s::Error::NotFound (GET /k8s/clusters/c-vkf6c/k8s/clusters/c-vkf6c/api/v1 => HTTP 404 Not Found: {"paths"=>["/apis", "/apis/", "/apis/apiextensions.k8s.io", "/apis/apiextensions.k8s.io/v1beta1", "/health
z", "/healthz/etcd", "/healthz/log", "/healthz/ping", "/healthz/poststarthook/generic-apiserver-start-informers", "/healthz/poststarthook/start-apiextensions-controllers", "/healthz/poststarthook/start-api
extensions-informers", "/metrics", "/openapi/v2", "/swagger-2.0.0.json", "/swagger-2.0.0.pb-v1", "/swagger-2.0.0.pb-v1.gz", "/swagger.json", "/swaggerapi", "/version"]})
```

Notice the `GET /k8s/clusters/c-vkf6c/k8s/clusters/c-vkf6c/api/v1`.

Using the "Maybe something like this" works better and all the tests still pass:

```
2.5.3 :001 > client = K8s::Client.config(K8s::Config.load_file(ENV['KUBECONFIG']))
 => #<K8s::Client:0x00007fd0520055a8 @transport=#<K8s::Transport:0x00007fd052005aa8 @server="https://REDACTED.example.com:443", @path_prefix="/k8s/clusters/c-vkf6c", @auth_token="...
2.5.3 :002 > client.api.resource('pods', namespace: 'cattle-system').list.each do |pod|
2.5.3 :003 >     puts "namespace=#{pod.metadata.namespace} pod: #{pod.metadata.name}"
2.5.3 :004?>   end
namespace=cattle-system pod: cattle-cluster-agent-78c9cd6ddb-wdf4c
namespace=cattle-system pod: cattle-node-agent-4b9zl
namespace=cattle-system pod: cattle-node-agent-88r5s
namespace=cattle-system pod: cattle-node-agent-8dqrd
...
```